tests/apps/python/numpy_check.py

```diff
@@ -0,0 +1,86 @@
+import reframe as rfm
+from reframe.core.decorators import run_after
+import reframe.utility.sanity as sn
+
+
+@rfm.simple_test
+class NumpyTest(rfm.RunOnlyRegressionTest):
+    def __init__(self):
+        self.descr = 'Test a few typical numpy operations'
+        self.valid_prog_environs = ['builtin']
+        self.modules = ['Python']
+        self.time_limit = '20m'
+
+        self.perf_patterns = {
+            'dot': sn.extractsingle(
+                r'^Dotted two \S* matrices in\s+(?P<dot>\S+)\s+s',
+                self.stdout, 'dot', float),
+            'svd': sn.extractsingle(
+                r'^SVD of a \S* matrix in\s+(?P<svd>\S+)\s+s',
+                self.stdout, 'svd', float),
+            'cholesky': sn.extractsingle(
+                r'^Cholesky decomposition of a \S* matrix in'
+                r'\s+(?P<cholesky>\S+)\s+s',
+                self.stdout, 'cholesky', float),
+            'eigendec': sn.extractsingle(
+                r'^Eigendecomposition of a \S* matrix in'
+                r'\s+(?P<eigendec>\S+)\s+s',
+                self.stdout, 'eigendec', float),
+            'inv': sn.extractsingle(
+                r'^Inversion of a \S* matrix in\s+(?P<inv>\S+)\s+s',
+                self.stdout, 'inv', float),
+        }
+
+        self.sanity_patterns = sn.assert_found(r'Numpy version:\s+\S+',
+                                               self.stdout)
+        self.executable = 'python3'
+        self.executable_opts = ['np_ops.py']
+        # self.use_multithreading = False
+        self.tags = {'python', 'antwerp', 'performance'}
+        self.maintainers = ['Lewih']
+        self.valid_systems = ['*:single-node']
+
+        self.reference = {
+            'vaughan:single-node': {
+                'dot': (0.30, None, 0.10, 'seconds'),
+                'svd': (0.55, None, 0.10, 'seconds'),
+                'cholesky': (0.3, None, 0.10, 'seconds'),
+                'eigendec': (7.0, None, 0.10, 'seconds'),
+                'inv': (0.40, None, 0.10, 'seconds'),
+            },
+            'genius:single-node': {
+                'dot': (0.46, None, 0.10, 'seconds'),
+                'svd': (0.67, None, 0.10, 'seconds'),
+                'cholesky': (0.15, None, 0.10, 'seconds'),
+                'eigendec': (7.0, None, 0.10, 'seconds'),
+                'inv': (0.22, None, 0.10, 'seconds'),
+            },
+            'hydra:single-node': {
+                'dot': (0.25, None, 0.10, 'seconds'),
+                'svd': (25.82, None, 0.10, 'seconds'),
+                'cholesky': (0.77, None, 0.10, 'seconds'),
+                'eigendec': (58.50, None, 0.10, 'seconds'),
+                'inv': (0.54, None, 0.10, 'seconds'),
+            },
+            'hortense:single-node': {
+                'dot': (1.28, None, 0.10, 'seconds'),
+                'svd': (7.54, None, 0.10, 'seconds'),
```

Comment:
I'm getting way better numbers when running `np_ops.py` in an interactive job on Hortense, so it looks like something wrong/weird is happening with the way in which ReFrame is submitting the job:

```
[vsc40023@node5059 ~]$ ml SciPy-bundle/2020.11-intel-2020b; export OMP_NUM_THREADS=16; export MKL_NUM_THREADS=16; python3 np_ops.py
Numpy version: 1.19.4
Dotted two 4096x4096 matrices in 0.23 s.
SVD of a 2048x1024 matrix in 0.36 s.
Cholesky decomposition of a 2048x2048 matrix in 0.20 s.
Eigendecomposition of a 2048x2048 matrix in 5.10 s.
Inversion of a 2048x2048 matrix in 0.25 s.
```

Reply:
I did some extra tests.
the problem is OpenBLAS, which is not able to handle more than 8 threads for this numpy test.
the optimal number of threads seems to be 6.
the same thing is happening with BLIS.
intel MKL does not have any problems with 16 threads.

these problems may also be dependent on exactly which cores are assigned to the job, as we don't have a single job per node policy.

```
ml SciPy-bundle/2021.10-foss-2021b; OMP_NUM_THREADS=6 python3 np_ops.py
Numpy version: 1.21.3
Dotted two 4096x4096 matrices in 0.42 s.
SVD of a 2048x1024 matrix in 0.64 s.
Cholesky decomposition of a 2048x2048 matrix in 0.14 s.
Eigendecomposition of a 2048x2048 matrix in 5.19 s.
Inversion of a 2048x2048 matrix in 0.23 s.

ml SciPy-bundle/2021.05-intel-2021a; MKL_NUM_THREADS=16 python3 np_ops.py 
Numpy version: 1.20.3
Dotted two 4096x4096 matrices in 0.44 s.
SVD of a 2048x1024 matrix in 0.47 s.
Cholesky decomposition of a 2048x2048 matrix in 0.15 s.
Eigendecomposition of a 2048x2048 matrix in 5.24 s.
Inversion of a 2048x2048 matrix in 0.43 s.
```

Reply:
> these problems may also be dependent on exactly which cores are assigned to the job, as we don't have a single job per node policy

I understood what is going on: Jobs are overlapping on Hortense, at least using `astaff` account.
By running ` srun -N 1 -A astaff --pty bash` and then `htop` this is the situation:
<img width="1440" alt="Screenshot 2022-05-01 at 16 17 00" src="https://user-images.githubusercontent.com/37453252/166150035-cd0666f0-04d5-448d-869a-d0438f97dcc4.png">
I would expect an entire node at my disposal, especially for testing purposes, but this is not the case.


In order to avoid such a phenomenon do `srun -N 1 -A astaff --exclusive --mem=0 --pty bash`.
<img width="1440" alt="Screenshot 2022-05-01 at 16 21 37" src="https://user-images.githubusercontent.com/37453252/166150160-bfb9f9c2-9163-4a92-be82-54f2af9b57f4.png">
Reference at this [link](https://slurm.schedmd.com/srun.html#OPT_exclusive).

Same goes for the batch file:
```
#!/bin/bash
#SBATCH --job-name="rfm_NumpyTest_job"
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=6
#SBATCH --output=rfm_NumpyTest_job.out
#SBATCH --error=rfm_NumpyTest_job.err
#SBATCH --time=0:20:0
#SBATCH --nodes=1
#SBATCH --exclusive
#SBATCH --mem=0
#SBATCH -A astaff
module load SciPy-bundle/2021.10-foss-2021b
export OMP_NUM_THREADS=6
export MKL_NUM_THREADS=6
python3 np_ops.py
```

With this configuration results are acceptable and job completes in 2 minutes with no issues on Hortense:
```
Numpy version: 1.21.3
Dotted two 4096x4096 matrices in 0.62 s.
SVD of a 2048x1024 matrix in 1.05 s.
Cholesky decomposition of a 2048x2048 matrix in 0.39 s.
Eigendecomposition of a 2048x2048 matrix in 9.60 s.
Inversion of a 2048x2048 matrix in 0.51 s.
```

Moreover, I tried installing eb stack for SciPy-bundle/2021.10-foss-2021b and SciPy-bundle/2021.10-intel-2021b on a virgin location on Vaughan: performance are lower with foss toolchain and it scales way worse over a high number of threads, but overall performance is not as bad as in an overlapping environment.

With 16 threads, as previously configured, job completes now within the first 3-4  minutes.
```
Numpy version: 1.21.3
Dotted two 4096x4096 matrices in 0.30 s.
SVD of a 2048x1024 matrix in 1.85 s.
Cholesky decomposition of a 2048x2048 matrix in 0.39 s.
Eigendecomposition of a 2048x2048 matrix in 14.03 s.
Inversion of a 2048x2048 matrix in 0.46 s.
```

I will test on hydra as well.